src/NumericalAlgorithms/LinearOperators/IndefiniteIntegral.cpp

```diff
@@ -6,62 +6,65 @@
+
+template <size_t Dim, size_t... Is>
+std::array<std::reference_wrapper<const Matrix>, Dim> make_integration_matrices(
+    const Mesh<Dim>& mesh, const size_t dim_to_integrate,
+    std::index_sequence<Is...> /*meta*/
+    ) noexcept {
+  return std::array<std::reference_wrapper<const Matrix>, Dim>{
+      {(Is == dim_to_integrate
+            ? Spectral::integration_matrix(mesh.slice_through(dim_to_integrate))
+            : empty_matrix)...}};
 }
 }  // namespace
 
-template <size_t Dim>
-void indefinite_integral(const gsl::not_null<DataVector*> integral,
-                         const DataVector& integrand, const Mesh<Dim>& mesh,
+template <size_t Dim, typename VectorType>
+void indefinite_integral(const gsl::not_null<VectorType*> integral,
+                         const VectorType& integrand, const Mesh<Dim>& mesh,
                          const size_t dim_to_integrate) noexcept {
-  if (UNLIKELY(integral->size() != integrand.size())) {
-    *integral = DataVector(integrand.size());
-  }
-  indefinite_integral_impl(integral, integrand, mesh, dim_to_integrate);
+  integral->destructive_resize(integrand.size());
+  apply_matrices(integral,
+                 make_integration_matrices<Dim>(
+                     mesh, dim_to_integrate, std::make_index_sequence<Dim>{}),
+                 integrand, mesh.extents());
 }
 
-template <size_t Dim>
-DataVector indefinite_integral(const DataVector& integrand,
+template <size_t Dim, typename VectorType>
+VectorType indefinite_integral(const VectorType& integrand,
                                const Mesh<Dim>& mesh,
                                const size_t dim_to_integrate) noexcept {
-  DataVector integral{};
-  indefinite_integral(&integral, integrand, mesh, dim_to_integrate);
+  VectorType integral{integrand.size()};
+  indefinite_integral(make_not_null(&integral), integrand, mesh,
+                      dim_to_integrate);
   return integral;
 }
 
 #define GET_DIM(data) BOOST_PP_TUPLE_ELEM(0, data)
+#define GET_VECTORTYPE(data) BOOST_PP_TUPLE_ELEM(1, data)
 
-#define INSTANTIATION(r, data)             \
-  template DataVector indefinite_integral( \
-      const DataVector&, const Mesh<GET_DIM(data)>&, const size_t) noexcept;
+#define INSTANTIATION(r, data)                                                \
+  template GET_VECTORTYPE(data)                                               \
+      indefinite_integral(const GET_VECTORTYPE(data)&,                        \
+                          const Mesh<GET_DIM(data)>&, const size_t) noexcept; \
+  template void indefinite_integral(                                          \
+      const gsl::not_null<GET_VECTORTYPE(data)*> integral,                    \
+      const GET_VECTORTYPE(data)&, const Mesh<GET_DIM(data)>&,                \
+      const size_t) noexcept;
 
-GENERATE_INSTANTIATIONS(INSTANTIATION, (1, 2, 3))
+GENERATE_INSTANTIATIONS(INSTANTIATION, (1, 2, 3),
+                        (DataVector, ComplexDataVector))
```

Comment:
You probably need to include `ComplexDataVector.hpp` (why does it compile without?)